### PR TITLE
ENT-3749: Squash ReplaceStr() and StringReplace()

### PR DESCRIPTION
**Pull request by karlhto**
From Jira:

```
We have two functions ReplaceStr() and StringReplace() that do pretty much the same, but not exactly. We should squash them and get rid of one of them - probably the former one as it has worse performance.

Acceptance criteria:
1. Only one function for replacing strings in place used in our codebase.
2. Unit tests for the function written and passing.
```

Currently WIP, still needs tests for `StringReplace()`

Merge together:
https://github.com/cfengine/enterprise/pull/483

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thanks for submitting a PR! Maybe @vpodzime can review this?

**Comment by karlhto** [comment]
Also, it is worth noting that this is blocked by https://github.com/cfengine/enterprise/pull/483

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics, my dude

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2035/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2035/

**Comment by karlhto** [comment]
@cf-bottom I forgot to add that I also need this PR: https://github.com/cfengine/enterprise/pull/483

Jenkins with exotics.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2036/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2036/

**Comment by karlhto** [comment]
@cf-bottom Try again with this PR: https://github.com/cfengine/enterprise/pull/483
Jenkins with exotics, my dude.

File: libutils/string_lib.c [hunk 1]
```diff
@@ -803,26 +803,26 @@ bool ReplaceStr(const char *in, char *out, int outSz, const char *from, const ch
 /**
  * Replace all occurrences of #find with #replace.
  *
```
Comment (vpodzime) [inline libutils/string_lib.c]: Extra comma in the commit message's title.

File: libutils/string_lib.c [hunk 2]
```diff
@@ -803,26 +803,26 @@ bool ReplaceStr(const char *in, char *out, int outSz, const char *from, const ch
...
+ssize_t StringReplace(char *buf, size_t buf_size,
+                      const char *find, const char *replace)
 {
     assert(find[0] != '\0');
 
     char *p = strstr(buf, find);
     if (p == NULL)
     {
-        return 0;
+        return strlen(buf);
```
Comment (vpodzime) [inline libutils/string_lib.c]: I think you are changing the *0 if no replace took place* part of the behavior here.

File: cf-agent/verify_packages.c [hunk 3]
```diff
@@ -3289,7 +3289,9 @@ bool ExecPackageCommand(EvalContext *ctx, char *command, int verify, int setCmdC
             }
         }
 
-        ReplaceStr(line, lineSafe, sizeof(lineSafe), "%", "%%");
+        char lineSafe[line_size * 2];
```
Comment (vpodzime) [inline cf-agent/verify_packages.c]: Shouldn't this be `2 * res` if we wanted to make it actually safe?

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2038/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2038/

File: tests/unit/string_lib_test.c [hunk 4]
```diff
@@ -351,27 +417,33 @@ static void test_string_to_long_errors(void)
...
+    assert_int_not_equal(StringToLong("56789d", &target), 0);
+    assert_int_equal(StringToLong("9999999999999999999999999999999",
+                                  &target),
+                     ERANGE);
+    assert_int_equal(StringToLong(" 999999999999999999999999999999",
+                                  &target),
+                     ERANGE);
+    assert_int_equal(StringToLong("-999999999999999999999999999999",
+                                  &target),
+                     ERANGE);
```
Comment (vpodzime) [inline tests/unit/string_lib_test.c]: These changes should go into a separate commit.

File: libutils/string_lib.c (hunk 2 above)
Comment (karlhto) [reply libutils/string_lib.c]: That was intentional. I think it's a more consistent behaviour in the case that you are using the result of the function as a way to gauge how long it is. But I also see that knowing if there was some replacement would be a good idea.

Should I change it back, or does my explanation make sense?

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics, please!

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2046/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2046/

File: libutils/string_lib.c (hunk 2 above)
Comment (karlhto) [reply libutils/string_lib.c]: Never mind. I'm being silly.

**Comment by karlhto** [comment]
@cf-bottom Let's try jenkins with exotics again.

Please use this PR https://github.com/cfengine/enterprise/pull/483

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2048/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2048/

File: libutils/string_lib.c (hunk 1 above)
Comment (karlhto) [reply libutils/string_lib.c]: Fixed

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics

cfengine/enterprise#483

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2049/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2049/

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics, please.

Merge with: https://github.com/cfengine/enterprise/pull/483

File: cf-agent/verify_packages.c (hunk 3 above)
Comment (karlhto) [reply cf-agent/verify_packages.c]: Yes, you are mostly correct, although it should be `(res + 1) * 2`, since `res` does not contain `\0`

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2064/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2064/

File: cf-agent/verify_packages.c (hunk 3 above)
Comment (karlhto) [reply cf-agent/verify_packages.c]: ...or `res * 2 + 1` is better, actually.

**Comment by karlhto** [comment]
@cf-bottom I think this is it, man. Jenkins with exotics.

Merge with: cfengine/enterprise#483

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2065/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2065/

**Comment by olehermanse** [comment]
@cf-bottom jenkins with exotics

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2080/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2080/
